### PR TITLE
Check whether build passes. Do not reuse builds across tests.

### PR DESCRIPTION
**Pull request by SlouchyButton**
Fixes: #444

This contains the same change as in ruby container PR https://github.com/sclorg/s2i-ruby-container/pull/556.

The other change is related to ruby container issue https://github.com/sclorg/s2i-ruby-container/issues/557.

Ruby container issue has a more thorough description on what's specifically the problem.

Reason why `docker kill` has to be changed to `docker stop` in `test/test-lib-nodejs.sh` L: 176 is that `docker kill` doesn't wait for the container to stop. It forcefully kills it and continues. This then creates a race condition with the clearing function. When the `docker rmi -f` is called, the previous kill call hasn't stopped the container fully yet, so the `docker rmi -f` will see the existing container. It will try to delete it, because we are forcing image deletion, but before it goes through the internal process the container gets stopped and deleted by the previous `docker kill` call.

When the `docker rmi -f` actually reaches the part where it wants to delete the container, it doesn't exist anymore and fails. `docker stop` on the other hand won't work asynchronously and will try to gracefully stop the container and in case it doesn't respond it will forcefully kill and continue only after the successful kill happened.

<!-- issue-commentator = {"comment-id":"2374228891"} -->











































































































































































































































<!-- testing-farm = {"lock":"false","comment-id":"2376604750","data":[{"id":"d1586df1-36d6-46e7-92a6-20c6fa86bcbb","name":"Fedora - 20-minimal","status":"complete","outcome":"passed","runTime":1174.339007,"created":"2024-09-30T10:01:51.354063","updated":"2024-09-30T10:01:51.354070","compose":"Fedora-latest","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.dev.testing-farm.io/d1586df1-36d6-46e7-92a6-20c6fa86bcbb\">test</a> ","<a href=\"https://artifacts.dev.testing-farm.io/d1586df1-36d6-46e7-92a6-20c6fa86bcbb/pipeline.log\">pipeline</a>"]},{"id":"bab7b7ff-0f33-46ef-a710-ebd58cfd5713","name":"Fedora - 22-minimal","status":"complete","outcome":"passed","runTime":621.186323,"created":"2024-09-30T10:02:15.966193","updated":"2024-09-30T10:02:15.966205","compose":"Fedora-latest","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.dev.testing-farm.io/bab7b7ff-0f33-46ef-a710-ebd58cfd5713\">test</a> ","<a href=\"https://artifacts.dev.testing-farm.io/bab7b7ff-0f33-46ef-a710-ebd58cfd5713/pipeline.log\">pipeline</a>"]},{"id":"ed067ada-7922-476d-9baf-6aedc089f5d6","name":"CentOS Stream 9 - 20-minimal","status":"complete","outcome":"passed","runTime":1203.200966,"created":"2024-09-30T10:01:48.056072","updated":"2024-09-30T10:01:48.056079","compose":"CentOS-Stream-9","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.dev.testing-farm.io/ed067ada-7922-476d-9baf-6aedc089f5d6\">test</a> ","<a href=\"https://artifacts.dev.testing-farm.io/ed067ada-7922-476d-9baf-6aedc089f5d6/pipeline.log\">pipeline</a>"]},{"id":"14df2a6a-18c3-4f2c-9f76-fd74a003c464","name":"Fedora - 18-minimal","status":"complete","outcome":"passed","runTime":640.364082,"created":"2024-09-30T10:01:51.068344","updated":"2024-09-30T10:01:51.068351","compose":"Fedora-latest","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.dev.testing-farm.io/14df2a6a-18c3-4f2c-9f76-fd74a003c464\">test</a> ","<a href=\"https://artifacts.dev.testing-farm.io/14df2a6a-18c3-4f2c-9f76-fd74a003c464/pipeline.log\">pipeline</a>"]},{"id":"48eb53d9-4e28-4309-9e96-7548086a09cd","name":"CentOS Stream 10 - 22-minimal","status":"complete","outcome":"passed","runTime":724.888216,"created":"2024-09-30T10:01:59.429188","updated":"2024-09-30T10:01:59.429195","compose":"CentOS-Stream-10","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.dev.testing-farm.io/48eb53d9-4e28-4309-9e96-7548086a09cd\">test</a> ","<a href=\"https://artifacts.dev.testing-farm.io/48eb53d9-4e28-4309-9e96-7548086a09cd/pipeline.log\">pipeline</a>"]},{"id":"14911b95-8584-4e4a-a63a-dcd966189522","name":"RHEL9 - 20-minimal","status":"complete","outcome":"passed","runTime":1597.534137,"created":"2024-09-30T10:01:49.855861","updated":"2024-09-30T10:01:49.855871","compose":"RHEL-9.4.0-Nightly","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.osci.redhat.com/testing-farm/14911b95-8584-4e4a-a63a-dcd966189522\">test</a> ","<a href=\"https://artifacts.osci.redhat.com/testing-farm/14911b95-8584-4e4a-a63a-dcd966189522/pipeline.log\">pipeline</a>"]},{"id":"31550683-8d9c-4e35-accd-7032f60bb6b4","name":"RHEL9 - 18-minimal","status":"complete","outcome":"passed","runTime":876.585201,"created":"2024-09-30T10:01:49.160727","updated":"2024-09-30T10:01:49.160736","compose":"RHEL-9.4.0-Nightly","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.osci.redhat.com/testing-farm/31550683-8d9c-4e35-accd-7032f60bb6b4\">test</a> ","<a href=\"https://artifacts.osci.redhat.com/testing-farm/31550683-8d9c-4e35-accd-7032f60bb6b4/pipeline.log\">pipeline</a>"]},{"id":"7780520f-c63e-4b5b-91c0-cff4cec799e7","name":"RHEL9 - 22-minimal","status":"complete","outcome":"passed","runTime":926.410167,"created":"2024-09-30T10:02:31.817618","updated":"2024-09-30T10:02:31.817626","compose":"RHEL-9.4.0-Nightly","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.osci.redhat.com/testing-farm/7780520f-c63e-4b5b-91c0-cff4cec799e7\">test</a> ","<a href=\"https://artifacts.osci.redhat.com/testing-farm/7780520f-c63e-4b5b-91c0-cff4cec799e7/pipeline.log\">pipeline</a>"]},{"id":"eb6c3969-26ea-4657-a455-caf4f263e7cc","name":"RHEL8 - 18-minimal","status":"complete","outcome":"passed","runTime":1192.418024,"created":"2024-09-30T10:01:48.664863","updated":"2024-09-30T10:01:48.664870","compose":"RHEL-8.10.0-Nightly","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.osci.redhat.com/testing-farm/eb6c3969-26ea-4657-a455-caf4f263e7cc\">test</a> ","<a href=\"https://artifacts.osci.redhat.com/testing-farm/eb6c3969-26ea-4657-a455-caf4f263e7cc/pipeline.log\">pipeline</a>"]},{"id":"3773025d-adc8-4585-98e6-98f29385c878","name":"RHEL8 - 20-minimal","runTime":1795.869833,"created":"2024-09-30T10:01:51.617232","updated":"2024-09-30T10:01:51.617240","compose":"RHEL-8.10.0-Nightly","arch":"x86_64","infrastructureFailure":false,"status":"complete","outcome":"passed","results":["<a href=\"https://artifacts.osci.redhat.com/testing-farm/3773025d-adc8-4585-98e6-98f29385c878\">test</a> ","<a href=\"https://artifacts.osci.redhat.com/testing-farm/3773025d-adc8-4585-98e6-98f29385c878/pipeline.log\">pipeline</a>"]},{"id":"a203f174-7572-41bd-ada6-8373cb7e25e4","name":"Fedora - 22","status":"complete","outcome":"passed","runTime":819.251808,"created":"2024-09-30T10:01:59.387077","updated":"2024-09-30T10:01:59.387088","compose":"Fedora-latest","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.dev.testing-farm.io/a203f174-7572-41bd-ada6-8373cb7e25e4\">test</a> ","<a href=\"https://artifacts.dev.testing-farm.io/a203f174-7572-41bd-ada6-8373cb7e25e4/pipeline.log\">pipeline</a>"]},{"id":"882ae27e-5892-4541-9345-be1216b27656","name":"CentOS Stream 10 - 22","status":"complete","outcome":"failed","runTime":797.179594,"created":"2024-09-30T10:02:01.084640","updated":"2024-09-30T10:02:01.084649","compose":"CentOS-Stream-10","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.dev.testing-farm.io/882ae27e-5892-4541-9345-be1216b27656\">test</a> ","<a href=\"https://artifacts.dev.testing-farm.io/882ae27e-5892-4541-9345-be1216b27656/pipeline.log\">pipeline</a>"]},{"id":"dd73da42-fa65-4d55-ba27-fbc563723e92","name":"CentOS Stream 9 - 20","status":"complete","outcome":"passed","runTime":1007.771423,"created":"2024-09-30T10:01:50.426281","updated":"2024-09-30T10:01:50.426289","compose":"CentOS-Stream-9","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.dev.testing-farm.io/dd73da42-fa65-4d55-ba27-fbc563723e92\">test</a> ","<a href=\"https://artifacts.dev.testing-farm.io/dd73da42-fa65-4d55-ba27-fbc563723e92/pipeline.log\">pipeline</a>"]},{"id":"1a854409-4569-45f8-b798-aa061a26e8ef","name":"Fedora - 18","status":"complete","outcome":"passed","runTime":899.800027,"created":"2024-09-30T10:01:46.939925","updated":"2024-09-30T10:01:46.939932","compose":"Fedora-latest","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.dev.testing-farm.io/1a854409-4569-45f8-b798-aa061a26e8ef\">test</a> ","<a href=\"https://artifacts.dev.testing-farm.io/1a854409-4569-45f8-b798-aa061a26e8ef/pipeline.log\">pipeline</a>"]},{"id":"138e08d4-8d64-4743-9fdf-074e384e3427","name":"RHEL9 - 18","status":"complete","outcome":"passed","runTime":1081.309967,"created":"2024-09-30T10:01:46.698241","updated":"2024-09-30T10:01:46.698249","compose":"RHEL-9.4.0-Nightly","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.osci.redhat.com/testing-farm/138e08d4-8d64-4743-9fdf-074e384e3427\">test</a> ","<a href=\"https://artifacts.osci.redhat.com/testing-farm/138e08d4-8d64-4743-9fdf-074e384e3427/pipeline.log\">pipeline</a>"]},{"id":"026f5d6e-0414-439c-ba1b-c4e6f8e30d1e","name":"RHEL9 - 20","status":"complete","outcome":"passed","runTime":1129.408314,"created":"2024-09-30T10:01:49.814596","updated":"2024-09-30T10:01:49.814604","compose":"RHEL-9.4.0-Nightly","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.osci.redhat.com/testing-farm/026f5d6e-0414-439c-ba1b-c4e6f8e30d1e\">test</a> ","<a href=\"https://artifacts.osci.redhat.com/testing-farm/026f5d6e-0414-439c-ba1b-c4e6f8e30d1e/pipeline.log\">pipeline</a>"]},{"id":"021fc666-b02c-446b-a00f-26c79eb49c39","name":"RHEL9 - 22","status":"complete","outcome":"passed","runTime":1186.598958,"created":"2024-09-30T10:02:03.509602","updated":"2024-09-30T10:02:03.509610","compose":"RHEL-9.4.0-Nightly","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.osci.redhat.com/testing-farm/021fc666-b02c-446b-a00f-26c79eb49c39\">test</a> ","<a href=\"https://artifacts.osci.redhat.com/testing-farm/021fc666-b02c-446b-a00f-26c79eb49c39/pipeline.log\">pipeline</a>"]},{"id":"a10d870b-83aa-4a2a-ac74-98466d901254","name":"RHEL8 - 20","status":"complete","outcome":"passed","runTime":1465.401497,"created":"2024-09-30T10:01:52.052999","updated":"2024-09-30T10:01:52.053006","compose":"RHEL-8.10.0-Nightly","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.osci.redhat.com/testing-farm/a10d870b-83aa-4a2a-ac74-98466d901254\">test</a> ","<a href=\"https://artifacts.osci.redhat.com/testing-farm/a10d870b-83aa-4a2a-ac74-98466d901254/pipeline.log\">pipeline</a>"]},{"id":"798fa90d-619c-4a46-b619-1187c608447a","name":"RHEL8 - 18","status":"complete","outcome":"passed","runTime":1250.871831,"created":"2024-09-30T10:01:46.941528","updated":"2024-09-30T10:01:46.941537","compose":"RHEL-8.10.0-Nightly","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.osci.redhat.com/testing-farm/798fa90d-619c-4a46-b619-1187c608447a\">test</a> ","<a href=\"https://artifacts.osci.redhat.com/testing-farm/798fa90d-619c-4a46-b619-1187c608447a/pipeline.log\">pipeline</a>"]},{"id":"797ff004-f0a1-4ccf-9935-d1dfc521e0c1","name":"Fedora - 20","status":"complete","outcome":"passed","runTime":936.549495,"created":"2024-09-30T10:01:50.380497","updated":"2024-09-30T10:01:50.380505","compose":"Fedora-latest","arch":"x86_64","infrastructureFailure":false,"results":["<a href=\"https://artifacts.dev.testing-farm.io/797ff004-f0a1-4ccf-9935-d1dfc521e0c1\">test</a> ","<a href=\"https://artifacts.dev.testing-farm.io/797ff004-f0a1-4ccf-9935-d1dfc521e0c1/pipeline.log\">pipeline</a>"]}]} -->

### CODE REVIEW

**Review by hhorak — CHANGES_REQUESTED** [review]
Please, include the PR description in the commit as well, github might not be here forever (if this reads somebody from GitHub, be aware that I wish it will be here forever, but you know... Microsoft... well, anyway...)

For the change itself, I do understand the `kill/stop` part, but I don't understand why you need to re-define `evaluate_build_result` function. I understand what it does, but when do we need to remove a container image when the build failed? Then nothing should be built, no? Or when to expect the build fails?

**Comment by hhorak** [comment]
[test]

**Comment by SlouchyButton** [comment]
> why you need to re-define `evaluate_build_result` function

That is related to this: https://github.com/sclorg/s2i-ruby-container/pull/556
I wanted both projects to use the same approach to detecting failed build.

The problem with `ct_check_testcase_result` is that it will happily continue and ignore that the build failed. It also won't add any entry to test result. It also (in case of the ruby container) didn't cause resulting return code of the test suite to be non-zero if build failed. You pointing it out made me take a look at that code again, and it needs a small change.

The thing that I am trying to achieve with that `evaluate_build_result` function, the tests will fail early and gracefully. If we rely only on `ct_check_testcase_result` that build might fail, and still continue and start a test on that build. In the ruby container, this resulted in a test reusing a built image from a previous test. Adding `evaluate_build_result` together with skipping the test when build failed added a new entry to the tests results that would inform you that the build failed.

It would fail early and not run any tests on that image, as it doesn't make sense to even try and run the tests on an image that failed to build. Otherwise, we would run the tests and fail on podman trying to start an image that doesn't exist for every test that depends on that image. It would work still, but I don't think it's a clean solution. Also, it helps with debugging as you will immediately see that the build failed, not the tests itself.

**Problem here is that I missed the part that should actually skip the test if the build failed, I will add that.**

**Comment by hhorak** [comment]
@SlouchyButton thanks for explaining, I think at least part of this explanation would be also worth keeping in the commit messages.

More thoughts:

- the problem with re-using image from previous test could be likely solved by removing the image (if it exists) before the build starts (in the build function directly) -- then we would end up with an image only when it was freshly built for the test case and the build succeeded; not sure whether this change would make the need of `evaluate_build_result` void, and `ct_check_testcase_result` could be used instead (with an improvement below)
- I like the idea to add more info about what steps of the test passed and what failed -- what if we add this option to the `ct_check_testcase_result` itself, so we could add such info in other places as well when we want (with some indentation, so we see the difference from the "final test result" and "sub-result")

**Comment by SlouchyButton** [comment]
> * what if we add this option to the `ct_check_testcase_result` itself, so we could add such info in other places as well

Love the idea, would definitely help with debugging as we would be able to pinpoint _exactly_ where the build failed. It would also allow us to continue with the 'fail early, fail gracefully' approach.

The thing is that since this is a test-lib function, we will have to go through every single container and check all uses of that function before changing it, as this change seems like it could be breaking.

Another thing I am a bit scared about is that we have already pushed Bash far beyond what it can handle in some places (https://github.com/sclorg/container-common-scripts/pull/380) and we are just adding more and more stuff onto it. This specifically shouldn't really be that complicated, but since we are changing test-lib, there is always a higher chance of breaking stuff, especially since this change would be modifying how that function works substantially.

I would still continue with this idea honestly, I like it, it will be a helpful feature to have. We should create a new issue for https://github.com/sclorg/container-common-scripts

**Comment by SlouchyButton** [comment]
> * solved by removing the image (if it exists) before the build starts

Currently, this PR removes the images after the test, so after merging that will be happening.

> not sure whether this change would make the need of evaluate_build_result void, and ct_check_testcase_result could be used instead (with an improvement below)

From how I understand that improvement we would basically make `ct_check_testcase_result` work similarly to how `evaluate_build_result` works now, so yes, there would be no need to have `evaluate_build_result` anymore.

**Comment by hhorak** [comment]
> The thing is that since this is a test-lib function, we will have to go through every single container and check all uses of that function before changing it, as this change seems like it could be breaking.

I don't think it would need to be breaking. The function `ct_check_testcase_result` currently accepts exactly one argument, so unless some container passes more arguments that have no meaning (that could be double-checked to be sure), we could make the second argument optional, so ignoring it would be legal and default. We'd only add the 2nd argument when we want extra info.

**Comment by SlouchyButton** [comment]
> I don't think it would need to be breaking.

The thing I had in mind is whether we could be modifying the final result with such subtest results everywhere.

After looking at this again I realized the `evaluate_build_result` is calling `ct_update_test_result` which actually adds it to the test log. That is also test-lib function, so this change shouldn't depend on specific container tests. I mistakenly thought that `evaluate_build_result` depends on a local test code. This shouldn't be as dramatic change as I made it seem.

**Comment by hhorak** [comment]
Cool. One more thing that bothers me is that the RHEL8 v18 tests fail repeatadly, while not seeing any nodejs failures in the nightly -- so I tried a different PR to see whether the failure is connected with this change or not: https://github.com/sclorg/s2i-nodejs-container/pull/446

**Comment by hhorak** [comment]
> RHEL8 v18 tests fail repeatadly,

Or is it possible that it failed silently before this fix? that could explain it...

**Comment by SlouchyButton** [comment]
> Or is it possible that it failed silently before this fix?

I don't think so. We added tests for builds failing, but the build passed. The only thing that failed was

`[FAILED] for 'hw' test_run_hw_application (00:00:11)`

```
fec7c64b7b9910932279506c33aca66a0f9af9dc2b63e762a29883cc77861801
time="2024-08-29T11:39:00-04:00" level=warning msg="StopSignal SIGTERM failed to stop container laughing_agnesi in 10 seconds, resorting to SIGKILL"
fec7c64b7b9910932279506c33aca66a0f9af9dc2b63e762a29883cc77861801
rm: cannot remove '/tmp/tmp.UzRE2OjMVt/./tmp.nh4Dei6dEd.cid': No such file or directory
Test for image 'ubi8/nodejs-18:1' FAILED (exit code: 1)
```
And it seems it failed due to the last command in the test function is `kill_test_application` and it's last command was `rm`. Normally, this command is used in other places so it _should_ work, but in this case the file wasn't found so it failed, and non-zero return code probably bubbled up through the functions as it was a last command.

The test function is interesting tho and I am not sure if it even does what it is supposed to do, since this code:
```
function test_run_hw_application() {
  # Verify that the HTTP connection can be established to test application container
  run_test_application hw
  # Wait for the container to write it's CID file
  wait_for_cid
  ct_check_testcase_result $?
  kill_test_application
}
```
AFAIK starts the test app, then waits for CID, but then when `ct_check_testcase_result` is called we are getting the resulting return code of `wait_for_cid`, which doesn't return anything, so I think we are effectively getting a return code of `sleep` in `wait_for_cid`. And then we kill it with no further testing.

Am I missing something here? @hhorak 

Furthermore, the test on rhel8 for version 18 passes locally for me.

**Comment by SlouchyButton** [comment]
[test]

File: test/test-lib-nodejs.sh [hunk 1]
```diff
@@ -609,5 +609,12 @@ function test_latest_imagestreams() {
   return $result
 }
 
+function cleanup() {
```
Comment (phracek) [inline test/test-lib-nodejs.sh:612]: What about to use https://github.com/sclorg/container-common-scripts/blob/master/test-lib.sh#L154 or this one https://github.com/sclorg/container-common-scripts/blob/master/test-lib.sh#L133?

**Review by phracek — CHANGES_REQUESTED** [review]
This pull request LGTM, but some nitpicks.

**Comment by phracek** [comment]
CentOS Stream 22 issue #452

File: test/test-lib-nodejs.sh (hunk 1 above)
Comment (SlouchyButton) [reply test/test-lib-nodejs.sh:612]: As far as I look both clean containers not images.

**Comment by phracek** [comment]
[test]

File: test/test-lib-nodejs.sh (hunk 1 above)
Comment (phracek) [reply test/test-lib-nodejs.sh:612]: I seem, then it makes sense

**Comment by phracek** [comment]
Rebase it please against master.

**Comment by SlouchyButton** [comment]
> Rebase it please against master.

@phracek done

**Comment by phracek** [comment]
[test]

**Comment by SlouchyButton** [comment]
In `test_run_binary_application`:
```
npm error In file included from rdkafka_int.h:116,
npm error                  from rdkafka.c:43:
npm error rdkafka_conf.h:40:10: fatal error: openssl/engine.h: No such file or directory
npm error    40 | #include <openssl/engine.h>
npm error       |          ^~~~~~~~~~~~~~~~~~
npm error compilation terminated.
```

It seems that the sample app uses OpenSSL engine library, but that one got deprecated/removed in F41. I'd guess that C10s already embraced this change. Test pass for Fedora as we are using F40 image that is not affected.

Ref: https://fedoraproject.org/wiki/Changes/OpensslDeprecateEngine

**Comment by phracek** [comment]
C10S issue is tracked here: #452

**Pull request merged**